### PR TITLE
[FLINK-14847][hive] Support retrieving Hive PK constraints

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

Get PK constraints for Hive tables.

## Brief change log

  - Add shim interface for retrieving PK constraints
  - Add PK to TableSchema when getting a table from HiveCatalog
  - Add test to verify.

## Verifying this change

New test case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
cc @godfreyhe @xuefuz @bowenli86 @zjuwangg

**Comment by wuchong** [comment]
We should hold this PR a little. @dawidwys proposed FLIP-87 to introduce primary key concept.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 1]
```diff
@@ -544,6 +544,12 @@ private CatalogBaseTable instantiateCatalogTable(Table hiveTable, HiveConf hiveC
 					properties,
 					comment);
 		} else {
+			List<String> primaryKey = client.getPrimaryKey(hiveTable.getDbName(), hiveTable.getTableName(),
+					HiveTableUtil.relyConstraint((byte) 0));
+			// Need a better place to hold the PK constraints. Put it as a property for now.
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: there is already a pr: [[FLINK-12846][table-common] Carry primary key information in TableSchema ](https://github.com/apache/flink/pull/10213) to solve this, add this info to the comments ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/config/CatalogConfig.java [hunk 2]
```diff
@@ -33,4 +33,9 @@
 	// Used to distinguish properties created by Hive and Flink,
 	// as Hive metastore has its own properties created upon table creation and migration between different versions of metastore.
 	public static final String FLINK_PROPERTY_PREFIX = "flink.";
+
+	/**
+	 * A comma separated list of columns that compose the primary key of a table.
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/config/CatalogConfig.java]: add some comments about "this is a temporary solution" ？

**Comment by lirui-apache** [comment]
Rebased and add Hive PK into TableSchema.
@godfreyhe @dawidwys @wuchong @KurtYoung  Please take a look. Thanks.

**Comment by KurtYoung** [comment]
@zjuwangg Please help reviewing codes from hive's side. I don't familiar with hive shim as well...

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV210.java [hunk 3]
```diff
@@ -55,4 +60,42 @@ public void alterPartition(IMetaStoreClient client, String databaseName, String
...
+				if (add) {
+					add = !HiveTableUtil.requireValidateConstraint(constraintTrait) || (boolean) isValidateMethod.invoke(constraint);
+				}
+				if (add) {
+					add = !HiveTableUtil.requireRelyConstraint(constraintTrait) || (boolean) isRelyMethod.invoke(constraint);
+				}
+				if (add) {
+					colNames.add((String) colNameMethod.invoke(constraint));
+				} else {
+					return null;
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV210.java]: minor: the signature is better to return an optional if value is nullable.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV210.java [hunk 4]
```diff
@@ -55,4 +60,42 @@ public void alterPartition(IMetaStoreClient client, String databaseName, String
...
+			Method isRelyMethod = constraintClz.getDeclaredMethod("isRely_cstr");
+			List<String> colNames = new ArrayList<>();
+			for (Object constraint : constraints) {
+				boolean add = !HiveTableUtil.requireEnableConstraint(constraintTrait) || (boolean) isEnableMethod.invoke(constraint);
+				if (add) {
+					add = !HiveTableUtil.requireValidateConstraint(constraintTrait) || (boolean) isValidateMethod.invoke(constraint);
+				}
+				if (add) {
+					add = !HiveTableUtil.requireRelyConstraint(constraintTrait) || (boolean) isRelyMethod.invoke(constraint);
+				}
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV210.java]: the logic here has been a bit confusing. can you add some explanation?

seems that `constaintTrait` can override `constraint`? if so, we can just process pk based on constraintTrait first; if constraintTrait is false, then process pk based on reflection?

Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV210.java]: `constaintTrait` is the desired traits that all PK columns should satisfy. If any trait is not satisfied by any PK column, we consider the PK as invalid and return null. For example, callers can specify they want a PK constraint that is both RELY and ENABLE. Therefore a constraint is only RELY wouldn't satisfy this requirement and null is returned.
I'll rename the parameters to make it easier to understand.

**Comment by lirui-apache** [comment]
@bowenli86 Updated to address your comments.

**Comment by bowenli86** [comment]
lgtm, will merge tomorrow

**Comment by KurtYoung** [comment]
Please rebase to master and I can help to merge this.

**Comment by lirui-apache** [comment]
@KurtYoung Just rebased. Please help with the merge. Thanks.

**Comment by lirui-apache** [comment]
My own travis build passed: https://travis-ci.org/lirui-apache/flink/builds/621525124

**Pull request merged**